File: Quickpaste/Services/PasteServices/PasteService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Quickpaste.Services.BlobServices;
using Quickpaste.Models;
using Quickpaste.Models.Dtos;
using Quickpaste.Data;
using AutoMapper;
using Microsoft.Extensions.Logging;
using Microsoft.AspNetCore.Http;

namespace Quickpaste.Services.PasteServices
{
    public class PasteService
    {
        private readonly IUploadLinkRepository _uploadLinkRepository;
        private readonly ILogger _logger;
        private readonly IPasteRepository _pasteRepository;
        private readonly BlobStorageService _storageService;


        public PasteService(BlobStorageService storageService, IPasteRepository pasteRepository, IUploadLinkRepository uploadLinkRepository, ILogger<PasteService> logger)
        {
            _storageService = storageService;
            _pasteRepository = pasteRepository;
            _logger = logger;
            _uploadLinkRepository = uploadLinkRepository;
        }

        /// <summary>
        /// Creates a new Paste given a PastePostDto
        /// </summary>
        /// <param name="pasteCreateModel"></param>
        /// <returns>The created Paste on success, otherwise null</returns>
        public async Task<Paste> CreatePaste(PastePostDto pasteCreateModel)
        {
            return await CreatePasteHelper(pasteCreateModel.QuickLink, pasteCreateModel.Message, pasteCreateModel.FileToUpload, pasteCreateModel.IsPublic);
            
        }

        /// <summary>
        /// Creates a paste based on an upload link
        /// </summary>
        /// <param name="uploadLink"></param>
        /// <param name="anonPasteCreateModel"></param>
        /// <returns>true on success, false on failure</returns>
        public async Task<bool> CreatePaste(UploadLink uploadLink, PasteAnonPostDto anonPasteCreateModel)
        {

            if (!uploadLink.AllowFile)
            {
                anonPasteCreateModel.FileToUpload = null;
            }

            Paste createdPaste = await CreatePasteHelper(uploadLink.QuickLink, anonPasteCreateModel.Message, anonPasteCreateModel.FileToUpload, false);
            if (createdPaste != null)
            {
                await _uploadLinkRepository.Delete(uploadLink.QuickLink);
                return true;
            }
            return false;
        }

        /// <summary>
        /// Helper method that actually creates a new paste and uploads a given file
        /// </summary>
        /// <param name="quickLink">The quick link associated with the paste</param>
        /// <param name="message">The paste's message</param>
        /// <param name="fileToUpload">The paste's file if it exists</param>
        /// <param name="isPublic">Whether the paste's content is publicly visible or not</param>
        /// <returns>Created Paste on success, null on failure</returns>
        private async Task<Paste> CreatePasteHelper(string quickLink, string message, IFormFile fileToUpload, bool isPublic)
        {

            Paste pasteToCreate = Paste.ToCreate(quickLink, message, null, isPublic);

            if (fileToUpload == null)
            {
                return await _pasteRepository.Create(pasteToCreate);
            }

            Paste createdPasteResult = null;
            string fileUrl = null;
            try
            {
                fileUrl = await _storageService.UploadFile(fileToUpload, pasteToCreate.IsPublic);

                if (fileUrl == null)
                {
                    return null;
                }

                pasteToCreate.BlobUrl = fileUrl;

                createdPasteResult = await _pasteRepository.Create(pasteToCreate);
            }
            finally
            {
                if (createdPasteResult == null && fileUrl != null) // if Paste could not be created but file was uploaded, delete the file
                {
                    bool successfulDelete = await _storageService.DeleteFile(fileUrl);
                    if (!successfulDelete)
                    {
                        _logger.LogWarning($"Could not delete blob for paste that couldn't be created: {fileUrl}");
                    }
                }
            }

            return createdPasteResult;
        }

        /// <summary>
        /// Deletes a paste and its associated blob if exists
        /// </summary>
        /// <param name="pasteId">ID of paste to delete</param>
        /// <returns>Deleted paste on success, null on failure</returns>
        public async Task<PasteGetDto> DeletePaste(string pasteId)
        {
            Paste pasteToDelete = await _pasteRepository.GetById(pasteId);

            if (pasteToDelete?.BlobUrl != null)
            {
                bool successfulBlobDelete = await _storageService.DeleteFile(pasteToDelete.BlobUrl);
                if (!successfulBlobDelete)
                {
                    return null;
                }

            }

            bool successfulDelete = await _pasteRepository.Delete(pasteId);
            return successfulDelete ? new PasteGetDto(pasteToDelete) : null;

        }

        /// <summary>
        /// Gets the paste with Id = pasteId
        /// </summary>
        /// <param name="pasteId"></param>
        /// <returns>Paste if found, null otherwise</returns>
        public async Task<PasteGetDto> GetPaste(string pasteId)
        {
            Paste paste = await _pasteRepository.GetById(pasteId);
            if (paste == null)
            {
                return null;
            }

            PasteGetDto pasteDto = new PasteGetDto(paste);
            return pasteDto;
        }

        /// <summary>
        /// Gets the paste with QuickLink = link
        /// </summary>
        /// <param name="link">The paste's quickLink</param>
        /// <returns>Paste if found, null otherwise</returns>
        public async Task<PasteGetDto> GetPasteByLink(string link)
        {
            Paste paste = await _pasteRepository.GetByLink(link);
            if (paste == null)
            {
                return null;
            }
            PasteGetDto pasteDto = new PasteGetDto(paste);
            return pasteDto;
        }

        /// <summary>
        /// Returns all pastes from latest to oldest
        /// </summary>
        /// <returns>List of Pastes</returns>
        public async Task<List<PasteGetDto>> GetAllPastes(int pageNumber = 1, int limit = 5)
        {
            limit = Math.Min(limit, 20);
            return (await _pasteRepository.GetAll(pageNumber, limit)).Select(r => { return new PasteGetDto(r); }).OrderByDescending(r => r.CreatedAt).ToList();
        }

        /// <summary>
        /// Create a link which can be used to view a blob that is normally private. After sharedLinkInput.durationInMinutes has passed, the link is no longer valid
        /// Also used to generate a temporary link for authenticated users to download their own blobs
        /// </summary>
        /// <param name="pasteId"></param>
        /// <param name="sharedLinkInput"></param>
        /// <returns>string URL which gives access to the normally private blob on success, otherwise null</returns>
        public async Task<SharedLinkGetDto> CreateSharedLink(string pasteId, SharedLinkPostDto sharedLinkInput)
        {
            Paste pasteToShare = await _pasteRepository.GetById(pasteId);
            if(pasteToShare == null)
            {
                return null;
            }
            string sharedLinkUrl = _storageService.GenerateSharedBlobUrl(pasteToShare.BlobUrl, sharedLinkInput.DurationInMinutes);
            if(sharedLinkUrl == null)
            {
                return null;
            }
            return new SharedLinkGetDto(sharedLinkInput.DurationInMinutes, sharedLinkUrl);
        }
    }
}

File: Quickpaste/Angular/src/app/pastes/paste.service.ts
import { Injectable } from "@angular/core";
import { HttpClient, HttpErrorResponse, HttpResponse } from "@angular/common/http";
import { Observable } from "rxjs/Observable";
import { PasteModel, PasteCreateModel, AnonPasteCreateModel } from "./models/paste.model";
import { AuthService } from "../auth/auth.service";
import { SharedLinkCreateModel, SharedLinkModel } from "./models/shared-link.model";
import "rxjs/add/operator/publishReplay";




@Injectable()
export class PasteService {

    /**
     * A map used to cache Pastes and retrieve them from memory
     */
    pasteMap: Map<string, Observable<PasteModel>> = new Map<string, Observable<PasteModel>>();

    constructor(private http: HttpClient, private authService: AuthService) { }

    /**
     * Gets the paste with the specified quickLink. The response is formatted into an array with a PasteModel and string for error message.
     * If the paste could not be retrieved, the 1st element of the array is set to null and the 2nd element contains an error message.
     * If the paste was found, the 1st element contains the PasteModel and the 2nd element is set to null
     * 
     * @param {string} quickLink 
     * @returns {Observable<[PasteModel?, string?]>} 
     * @memberof PasteService
     */
    getPaste(quickLink: string): Observable<[PasteModel, string]> {
        let cachedPaste = this.pasteMap.get(quickLink)
        if (cachedPaste) {
            return cachedPaste.map((resp: PasteModel) => {
                let formattedResponse: [PasteModel, string] = [resp, null];
                return formattedResponse;
            });
        }
        return this.http.get<PasteModel>(`/api/pastes/${quickLink}?isLink=true`, { headers: this.authService.authHeader() })
            .map((paste: PasteModel) => {
                this.pasteMap.set(paste.quick_link, Observable.of(paste));
                let formattedResponse: [PasteModel, string] = [paste, null];
                return formattedResponse;
            })
            .catch((error: HttpErrorResponse) => {
                let formattedResponse: [PasteModel, string] = [null, 'An unexpected error occured'];

                if (error.status === 404) {
                    formattedResponse[1] = 'Paste not found';
                } else if (error.status === 401) {
                    formattedResponse[1] = 'This paste is private';
                }
                return Observable.of(formattedResponse);
            });

    }

    /**
     * Get the list of all paste models, returning an array of PasteModels on success, or null on failure
     * 
     * @param {number} [pageNumber=1]  
     * @param {number} [limit=10]  
     * @returns {Observable<PasteModel[]>} 
     * @memberof PasteService
     */
    getPastes(pageNumber: number = 1, limit: number = 10): Observable<PasteModel[]> {

        return this.http.get<PasteModel[]>(`/api/pastes?pageNumber=${pageNumber}&limit=${limit}`, { headers: this.authService.authHeader() })
            .do((pastes: PasteModel[]) => {
                pastes.forEach(paste => {
                    this.pasteMap.set(paste.quick_link, Observable.of(paste));
                });
            });

    }

    /**
     * Create a new paste and return the created PasteModel on success, or null on failure
     * 
     * @param {PasteCreateModel} pasteToCreate 
     * @returns {Observable<PasteModel?>}
     * @memberof PasteService
     */
    createPaste(pasteToCreate: PasteCreateModel): Observable<PasteModel> {
        let formData = new FormData();
        if (pasteToCreate.file_to_upload !== null) {
            formData.append('file_to_upload', pasteToCreate.file_to_upload[0]);
        }
        formData.append('is_public', pasteToCreate.is_public + "");
        formData.append('message', pasteToCreate.message);
        formData.append('quick_link', pasteToCreate.quick_link);

        return this.http.post<PasteModel>("api/pastes", formData, { headers: this.authService.authHeader() })
            .do((paste: PasteModel) => this.pasteMap.set(paste.quick_link, Observable.of(paste)))
            .publishReplay(1)
            .refCount()
            .catch((error: HttpErrorResponse) => {
                console.log(error.status);
                console.log(error.error);
                return Observable.of(null);
            });
    }

    /**
     * Generates a random alphanumeric string
     * 
     * @returns {string} 
     * @memberof PasteService
     */
    generateRandomPasteLink(): string {
        let link = "";
        let charset = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
        let length = 32;

        for (let i = 0; i < length; i++) {
            link += charset.charAt(Math.floor(Math.random() * charset.length));
        }
        return link;
    }

    /**
     *  Create a link which can be used to view a blob for the paste that is normally private. After durationInMinutes has passed, the link is no longer valid
     * 
     * @param {string} pasteId 
     * @param {number} durationInMinutes 
     * @returns {Observable<SharedLinkModel>} 
     * @memberof PasteService
     */
    createSharedLink(pasteId: string, durationInMinutes: number): Observable<SharedLinkModel> {
        let sharedLinkInput: SharedLinkCreateModel = { duration_in_minutes: durationInMinutes };

        return this.http.post<SharedLinkModel>(`/api/pastes/${pasteId}/sharedlinks`, sharedLinkInput, { headers: this.authService.authHeader() })
            .catch((error: HttpErrorResponse) => {
                console.log(error.status);
                console.log(error.error);
                return Observable.of(null);
            });
    }

    /**
     * Create a paste anonymously using an upload link
     * 
     * @param {string} quickLink 
     * @param {AnonPasteCreateModel} pasteToCreate 
     * @returns {Observable<boolean>} 
     * @memberof PasteService
     */
    createPasteFromLink(quickLink: string, pasteToCreate: AnonPasteCreateModel): Observable<boolean> {
        let formData = new FormData();
        if (pasteToCreate.file_to_upload !== null) {
            formData.append('file_to_upload', pasteToCreate.file_to_upload[0]);
        }
        formData.append('message', pasteToCreate.message);

        return this.http.post(`api/pastes/${quickLink}`, formData, { headers: this.authService.authHeader() })
            .map((resp: any) => {
                return true;
            })
            .publishReplay(1)
            .refCount()
            .catch((error: HttpErrorResponse) => {
                console.log(error.status);
                console.log(error.error);
                return Observable.of(false);
            });
    }


    /**
     * Deletes a paste with the given id
     * 
     * @param {string} id 
     * @returns {Observable<boolean>} 
     * @memberof PasteService
     */
    deletePaste(id: string): Observable<boolean> {

        return this.http.delete(`/api/pastes/${id}`, { headers: this.authService.authHeader() })
            .map((deletedPaste: PasteModel) => {
                this.pasteMap.delete(deletedPaste.quick_link);
                return true;
            })
            .catch((error: HttpErrorResponse) => {
                return Observable.of(false);
            });

    }

}
